packages/hardhat-core/src/internal/artifacts.ts

```diff
@@ -867,6 +886,235 @@ Please replace "${contractName}" for the correct contract name wherever you are
   }
 }
 
+export class Artifacts implements IArtifacts {
+  private readonly _sourcesInPriorityOrder: [
+    IHardhatArtifactSource,
+    ...ArtifactSource[]
+  ];
+
+  constructor(artifactsPath: string) {
+    this._sourcesInPriorityOrder = [new HardhatArtifactSource(artifactsPath)];
+  }
+
+  public async readArtifact(
+    contractNameOrFullyQualifiedName: string
+  ): Promise<Artifact> {
+    const artifact = this._getFirstValueFromSources(
+      "readArtifact",
+      contractNameOrFullyQualifiedName
+    );
```

Comment:
In general I would avoid doing this sort of abstraction, especially if there are only three methods. This is a personal preference (strongish one though), but it would also be more consistent with the rest of the codebase, were we prefer repetition over meta-programming.

Reply:
Yeah, I'm not super happy with it, but it did seem to me to be better than repeating that whole try/catch block and all that mess yet again. I'm already repeating that mess twice (once for sync and once for async), and I've already had headaches from changing one but forgetting about the other, and I really wanted to avoid having 5 copies of that method instead of 2. But as usual, I'm happy to defer to convention and to your wisdom.